File: backend/src/homelab_cmd/api/schemas/containers.py
"""Pydantic schemas for Docker container API endpoints (US0158 - EP0014)."""

from datetime import datetime

from pydantic import BaseModel, Field


class ContainerInfo(BaseModel):
    """Individual Docker container information."""

    id: str = Field(..., description="Short container ID (12 chars)")
    name: str = Field(..., description="Container name")
    image: str = Field(..., description="Image name with tag")
    state: str = Field(
        ...,
        description="Container state: running, exited, created, paused, restarting, dead",
    )
    status: str = Field(..., description="Human-readable status from Docker")
    ports: str | None = Field(None, description="Port mappings")
    created_at: str | None = Field(None, description="Container creation timestamp")
    uptime_seconds: int | None = Field(
        None, description="Uptime in seconds (only for running containers)"
    )


class ContainerListResponse(BaseModel):
    """Response for container listing endpoint (US0158)."""

    server_id: str = Field(..., description="Server identifier")
    containers: list[ContainerInfo] = Field(
        default_factory=list, description="List of containers"
    )
    total: int = Field(..., description="Total container count")
    cached: bool = Field(False, description="Whether response was served from cache")
    fetched_at: datetime = Field(..., description="When data was fetched/cached")
    error: str | None = Field(None, description="Error message if fetch failed")


class ContainerActionResponse(BaseModel):
    """Response for container action endpoints (US0160-US0162)."""

    success: bool = Field(..., description="Whether the action succeeded (exit_code == 0)")
    output: str | None = Field(None, description="Docker command output (stdout or stderr)")
    container_id: str = Field(..., description="Container ID or name that was acted upon")
    action: str = Field(..., description="Action performed: start, stop, restart")

File: frontend/src/api/containers.ts
/**
 * Container API client (US0159, US0160, US0161 - EP0014).
 */

import { api } from './client';
import type { ContainerActionResponse, ContainerListResponse } from '../types/container';

/**
 * List Docker containers on a server.
 *
 * @param serverId - Server identifier
 * @param refresh - Force refresh, bypassing cache
 * @returns Container list response
 */
export async function getContainers(
  serverId: string,
  refresh = false
): Promise<ContainerListResponse> {
  const params = new URLSearchParams();
  if (refresh) {
    params.set('refresh', 'true');
  }
  const query = params.toString();
  const url = `/api/v1/servers/${serverId}/containers${query ? `?${query}` : ''}`;
  return api.get<ContainerListResponse>(url);
}

/**
 * Start a stopped Docker container (US0160).
 *
 * @param serverId - Server identifier
 * @param containerId - Container ID or name to start
 * @returns Container action response
 */
export async function startContainer(
  serverId: string,
  containerId: string
): Promise<ContainerActionResponse> {
  return api.post<ContainerActionResponse>(
    `/api/v1/servers/${serverId}/containers/${containerId}/start`
  );
}

/**
 * Stop a running Docker container (US0161).
 *
 * @param serverId - Server identifier
 * @param containerId - Container ID or name to stop
 * @param timeout - Graceful shutdown timeout in seconds (default 10)
 * @returns Container action response
 */
export async function stopContainer(
  serverId: string,
  containerId: string,
  timeout = 10
): Promise<ContainerActionResponse> {
  const params = new URLSearchParams();
  if (timeout !== 10) {
    params.set('timeout', String(timeout));
  }
  const query = params.toString();
  return api.post<ContainerActionResponse>(
    `/api/v1/servers/${serverId}/containers/${containerId}/stop${query ? `?${query}` : ''}`
  );
}

/**
 * Restart a Docker container (US0162).
 *
 * @param serverId - Server identifier
 * @param containerId - Container ID or name to restart
 * @returns Container action response
 */
export async function restartContainer(
  serverId: string,
  containerId: string
): Promise<ContainerActionResponse> {
  return api.post<ContainerActionResponse>(
    `/api/v1/servers/${serverId}/containers/${containerId}/restart`
  );
}
